src/lib/temple/back/evm-dapp.ts

```diff
@@ -180,6 +192,197 @@ export const connectEvm = async (origin: string, chainId: string, icon?: string)
+
+      return;
+    }
+
+    const rpcUrl = params.rpcUrls.find(url => url.startsWith('https'));
+
+    if (!rpcUrl) {
+      reject(
+        new ErrorWithCode(
+          EVMErrorCodes.INVALID_PARAMS,
+          `Expected array with at least one valid string HTTPS URL 'rpcUrls'. Received: ${params.rpcUrls}`
+        )
+      );
+
+      return;
+    }
+
+    const chainMetadata: EvmChainToAddMetadata = {
+      chainId: params.chainId,
+      name: params.chainName,
+      nativeCurrency: params.nativeCurrency,
+      rpcUrl,
+      blockExplorerUrl: params.blockExplorerUrls?.at(0)
+    };
+
+    await requestConfirm({
+      id,
+      payload: {
+        chainType: TempleChainKind.EVM,
+        type: 'add_chain',
+        metadata: chainMetadata,
+        chainId: currentChainId,
+        origin,
+        appMeta: dApp.appMeta
+      },
+      onDecline: () => {
+        reject(new ErrorWithCode(EVMErrorCodes.USER_REJECTED_REQUEST, 'Chain adding declined'));
+      },
+      handleIntercomRequest: async (confirmReq, decline) => {
+        if (confirmReq?.type === TempleMessageType.DAppAddEvmChainRequest && confirmReq.id === id) {
+          if (confirmReq.confirmed) {
+            await withUnlocked(async ({ vault }) => {
+              const chainIdNum = Number(chainMetadata.chainId);
+
+              const prevStoredSpecs = await fetchFromStorage(EVM_CHAINS_SPECS_STORAGE_KEY);
+              const prevSpecsWithFallback = prevStoredSpecs ?? {};
+
+              if (!prevSpecsWithFallback[chainIdNum] && !DEFAULT_EVM_CHAINS_SPECS[chainIdNum]) {
+                const newChainSpec: EvmChainSpecs = {
+                  name: chainMetadata.name,
+                  currency: {
+                    address: EVM_TOKEN_SLUG,
+                    standard: EvmAssetStandard.NATIVE,
+                    ...chainMetadata.nativeCurrency
+                  },
+                  testnet: confirmReq.testnet
+                };
+
+                await putToStorage(EVM_CHAINS_SPECS_STORAGE_KEY, {
+                  ...prevSpecsWithFallback,
+                  [chainIdNum]: newChainSpec
+                });
+              }
+
+              const settings = await vault.fetchSettings();
+              const customEvmNetworks = settings.customEvmNetworks ?? [];
+
+              if (
+                !customEvmNetworks.some(n => n.rpcBaseURL === rpcUrl) &&
+                !EVM_DEFAULT_NETWORKS.some(n => n.rpcBaseURL === rpcUrl)
+              ) {
+                console.log(2);
+                const updatedSettings = await vault.updateSettings({
+                  customEvmNetworks: [
+                    ...customEvmNetworks,
+                    {
+                      id,
+                      name: chainMetadata.name,
+                      chain: TempleChainKind.EVM,
+                      chainId: chainIdNum,
+                      color: COLORS[Math.floor(Math.random() * COLORS.length)],
```

Comment:
The same logic for color generation is already in `src/app/pages/ChainSettings/use-chain-operations.ts` and `src/app/templates/NetworksSettings/add-network-modal/use-add-network.ts`. Please consider creating a special function in `src/lib/ui/colors.ts` and using it